### PR TITLE
Rework partitioning

### PR DESCRIPTION
**Pull request by mschroederi**
**Highlights**
- Input data are now partitioned to the different SampleWorkers
- Server only sends weights back to those partitions it got a GradientMessage form

**Minor Changes**
- Additionally support for Confluent's Kafka Cluster UI was added
- Readme in dev folder

### CODE REVIEW

File: src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java [hunk 1]
```diff
@@ -2,41 +2,70 @@
...
+
+            // Remove first element of ring buffer
+            // if the buffer will exceed the maximum allowed buffer size
+            if (data.value.size() >= this.bufferSize) {
+                data.value.remove(0);
+            }
+
+            // Add newest data record to buffer
+            data.value.add(inputDataMessage.getValues());
+            this.inputDataBuffer.put(data.key, data.value);
```
Comment (rschwanhold) [inline src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java:57]: Is `data.key` the same as `partitionKey`?

File: src/main/java/de/hpi/datastreams/processors/WorkerTrainingProcessor.java [hunk 2]
```diff
@@ -48,26 +50,59 @@ public void init(ProcessorContext context) {
...
+        }
+
+        // TODO: remove
         try {
             Thread.sleep(1000);
         } catch (InterruptedException e) {
             e.printStackTrace();
         }
 
+        // Extract weights from message and store in local HashMap
```
Comment (rschwanhold) [inline src/main/java/de/hpi/datastreams/processors/WorkerTrainingProcessor.java]: This is skipped when there is no training data. Is this a problem or do we assume that this will only be the case at the beginning because once there is data, it will never be empty again?

File: src/main/java/de/hpi/datastreams/apps/BaseKafkaApp.java [hunk 3]
```diff
@@ -48,6 +49,12 @@ private Properties getProperties() {
         props.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, JSONSerde.class.getName());
         props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
 
+        props.put(StreamsConfig.PRODUCER_PREFIX + ProducerConfig.INTERCEPTOR_CLASSES_CONFIG,
+                "io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor");
+        props.put(
+                StreamsConfig.CONSUMER_PREFIX + ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG,
+                "io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor");
+
```
Comment (rschwanhold) [inline src/main/java/de/hpi/datastreams/apps/BaseKafkaApp.java:58]: I am not really understanding what this does. Maybe you can explain it to me.

File: src/main/java/de/hpi/datastreams/processors/ServerProcessor.java [hunk 4]
```diff
@@ -41,28 +41,27 @@ public void init(ProcessorContext context) {
...
         });
 
-        // TODO: start training iteration - sequential consistency model
+        // Start new training iteration on workers by sending update of weights
+        // TODO: Currently only eventual consistency model supported
         KeyRange keyRange = this.getKeyRange();
         WeightsMessage weights = new WeightsMessage(message.getVectorClock() + 1, keyRange, this.getWeights(keyRange));
 
-        this.weightsMessageProducer.send(new ProducerRecord<>(WEIGHTS_TOPIC, 0L, weights));
+        this.weightsMessageProducer.send(new ProducerRecord<>(WEIGHTS_TOPIC, message.getPartitionKey(), weights));
```
Comment (rschwanhold) [inline src/main/java/de/hpi/datastreams/processors/ServerProcessor.java:64]: Is this the point where we rely on the fact that we have four servers, but we only want one in the future?

File: src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java [hunk 5]
```diff
@@ -2,41 +2,70 @@
...
 import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
 import org.apache.kafka.streams.state.KeyValueStore;
 
+import java.util.Map;
+
 public class WorkerSamplingProcessor extends AbstractProcessor<Long, DataMessage> {
 
-    private KeyValueStore<Long, DataMessage> inputDataBuffer;
+    private KeyValueStore<Long, MyArrayList<Map<Integer, Float>>> inputDataBuffer;
```
Comment (rschwanhold) [inline src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java:16]: Previously we had a `DataMessage` as value. Wouldn't this mean that we now want a `MyArrayList<DataMessage>` as value?

**Review by rschwanhold — APPROVED** [review]
I have added a few comments. If you think they are all handled, then feel free to merge the PR.

File: src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java (hunk 5 above)
Comment (mschroederi) [reply src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java:16]: As a DataMessage contains several information that we do not need (like `vectorClock` or `keyRange`) we just  have to keep the essential information (`values`) of the DataMessage. Keep in mind here, that the training data is split horizontally so the `keyRange` always contains all feature ids.
So we can simply reduce memory consumption by just storing `key -> value` pairs of the input data.

File: src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java (hunk 1 above)
Comment (mschroederi) [reply src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java:57]: Yes, it is. `data` is a `KeyValue` we get from the iterator which iterates over `this.inputDataBuffer`.

File: src/main/java/de/hpi/datastreams/processors/WorkerTrainingProcessor.java (hunk 2 above)
Comment (mschroederi) [reply src/main/java/de/hpi/datastreams/processors/WorkerTrainingProcessor.java]: Good point. We might want to consider updating the weights anyway as the server will not send them again until it receives a `GradientMessage` from the specific `partitionKey`.

Comment (mschroederi) [reply src/main/java/de/hpi/datastreams/processors/WorkerTrainingProcessor.java]: Moreover, we should think of the fact that the loop will break unless we send a `GradientMessage` every time we receive a `WeightsMessage` from the server..

File: src/main/java/de/hpi/datastreams/apps/BaseKafkaApp.java (hunk 3 above)
Comment (mschroederi) [reply src/main/java/de/hpi/datastreams/apps/BaseKafkaApp.java:58]: This is needed to be able to track the actual message flow within Confluent's ControlCenter

File: src/main/java/de/hpi/datastreams/processors/ServerProcessor.java (hunk 4 above)
Comment (mschroederi) [reply src/main/java/de/hpi/datastreams/processors/ServerProcessor.java:64]: I had to insert a `partitionKey` attribute within the `GradientMessage` so we can make sure that we send back the `WeightsMessage` to the worker we received the `GradientMessage` from.
Keep in mind here that `partitionKey` in `ServerProcessor.process` will always be `0` due to the fact that all workers send their gradients to partitionKey `0L`.
We might have to rethink that concept in the future due to heavy load on the one `ServerProcessor`.
